### PR TITLE
Optimize Enum.map/2 for ranges

### PR DESCRIPTION
**Pull request by sabiwara**
Adding a specialized implementation of `Enum.map/2` for ranges leads to a noticeable speed improvement ([~1.40x](https://github.com/sabiwara/elixir_benches/commit/14c8e3a7b327591e3eaa6fb708758d371223c427)).

- the implementation is based on the existing [`Enum.reduce/3`](https://github.com/elixir-lang/elixir/blob/89e3ecc944f2eb5508e4614c3a1e3279086bdbea/lib/elixir/lib/enum.ex#L4348-L4356) implementation
- when benchmarking, I noticed that inlining the private helpers made a huge difference and could also speed-up `reduce_range/5`, so I added it as well
- this will also impact `for` which compiles to `Enum.map/2` in some cases since 1.15

### CODE REVIEW

File: lib/elixir/test/elixir/enum_test.exs [hunk 1]
```diff
@@ -1861,6 +1861,7 @@ defmodule EnumTest.Range do
   test "map/2" do
     assert Enum.map(1..3, fn x -> x * 2 end) == [2, 4, 6]
     assert Enum.map(-1..-3, fn x -> x * 2 end) == [-2, -4, -6]
+    assert Enum.map(1..10//2, fn x -> x * 2 end) == [2, 6, 10, 14, 18]
```
Comment (ericmj) [inline lib/elixir/test/elixir/enum_test.exs:1864]: Can you add tests for negative steps and the cases that return empty lists?

**Review by ericmj — APPROVED** [review]
Looks good. Just need some more tests to exercise the guards more.

File: lib/elixir/test/elixir/enum_test.exs (hunk 1 above)
Comment (sabiwara) [reply lib/elixir/test/elixir/enum_test.exs:1864]: Great point, added in [88d85c9](https://github.com/elixir-lang/elixir/pull/12554/commits/88d85c9ec7e0300e47d0d7861c79ed186edf9870).

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
